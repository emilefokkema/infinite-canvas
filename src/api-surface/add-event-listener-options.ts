/**
 * Options that can be provided when adding an event listener to an {@link InfiniteCanvas}
 */
export interface AddEventListenerOptions{
    /**
     * If `true`, the listener will be invoked at most once
     */
    once: boolean;
}