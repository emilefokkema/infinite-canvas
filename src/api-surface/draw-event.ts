import { Transformed } from "./transformed";

export interface DrawEvent extends Transformed{
}
