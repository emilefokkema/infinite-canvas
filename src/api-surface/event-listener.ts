import { EventMap } from "./event-map";


export declare type EventListener<K extends keyof EventMap> = (ev: EventMap[K]) => void;
