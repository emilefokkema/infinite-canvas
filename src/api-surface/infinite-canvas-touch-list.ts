import { InfiniteCanvasTouch } from "./infinite-canvas-touch";

export interface InfiniteCanvasTouchList extends TouchList{
    item(index: number): InfiniteCanvasTouch | null;
    [index: number]: InfiniteCanvasTouch;
}

export * from './infinite-canvas-touch';