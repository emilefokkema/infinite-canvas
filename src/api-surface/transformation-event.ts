import { Transformed } from "./transformed";

export interface TransformationEvent extends Event, Transformed{
}
