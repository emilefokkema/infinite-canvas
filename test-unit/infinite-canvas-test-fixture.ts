import { Mock, MockInstance, vi } from 'vitest'
import { DrawingLock } from "src/drawing-lock";
import { MockCanvasMeasurementProvider } from './mock-canvas-measurement-provider';
import { CanvasContextMock } from './canvas-context-mock';
import { Config } from 'api/config';
import { MockDrawingIterationProvider } from './mock-drawing-iteration-provider';
import { CssLengthConverterFactory } from 'src/css-length-converter-factory';
import { InfiniteCanvasViewBox } from 'src/infinite-canvas-viewbox';
import { RectangleManagerImpl } from 'src/rectangle/rectangle-manager-impl';
import { InfiniteContext } from 'src/infinite-context/infinite-context';
import { ViewBox } from 'src/interfaces/viewbox';
import { InfiniteCanvasRenderingContext2D } from 'api/infinite-canvas-rendering-context-2d';


function setupGlobals(){
	window.createImageBitmap = function(): Promise<ImageBitmap>{return undefined;};
	(<any>window).ImageData = class {
		public width: number;
		public height: number;
		public data: Uint8ClampedArray;
		constructor(arrayOrWidth: Uint8ClampedArray | number, widthOrHeight: number, height?: number){
			if(typeof arrayOrWidth === "number"){
				this.width = arrayOrWidth;
				this.height = widthOrHeight;
			}else{
				this.width = widthOrHeight;
				this.height = height;
				this.data = arrayOrWidth;
			}
		}
	};
}

export interface InfiniteCanvasTestFixture{
    width: number
    height: number
    releaseDrawingLockSpy: MockInstance<()=>void>
    getDrawingLockSpy: MockInstance<()=>DrawingLock>
    measurementProvider: MockCanvasMeasurementProvider
    contextMock: CanvasContextMock
    config: Partial<Config>
    mockDrawingIterationProvider: MockDrawingIterationProvider
    viewbox: ViewBox,
    infiniteContext: InfiniteCanvasRenderingContext2D
}

export function createInfiniteCanvasTestFixture(): InfiniteCanvasTestFixture{
    setupGlobals();
    const drawingLock: DrawingLock = {release(){}};
    const releaseDrawingLockSpy = vi.spyOn(drawingLock, 'release');
    const getDrawingLockSpy = vi.fn<()=>DrawingLock>().mockReturnValue(drawingLock)
    const width = 200;
    const height = 200;
    const measurementProvider = new MockCanvasMeasurementProvider(width, height);
    const contextMock = new CanvasContextMock();
    const config = {}
    const mockDrawingIterationProvider = new MockDrawingIterationProvider()
    const cssLengthConverterFactory: CssLengthConverterFactory = {
        create: () => ({
            getNumberOfPixels(value: number){
                return value;
            }
        })
    };
    const viewbox = new InfiniteCanvasViewBox(
        new RectangleManagerImpl(measurementProvider, config),
        contextMock.mock,
        mockDrawingIterationProvider,
        getDrawingLockSpy,
        () => false);
    const infiniteContext = new InfiniteContext(undefined, viewbox, cssLengthConverterFactory);
    return {
        width,
        height,
        releaseDrawingLockSpy,
        getDrawingLockSpy,
        measurementProvider,
        contextMock,
        config,
        mockDrawingIterationProvider,
        viewbox,
        infiniteContext
    };
}