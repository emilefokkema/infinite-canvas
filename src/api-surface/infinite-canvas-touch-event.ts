import { InfiniteCanvasTouchList } from "./infinite-canvas-touch-list";

export interface InfiniteCanvasTouchEvent extends TouchEvent{
    readonly targetTouches: InfiniteCanvasTouchList;
    readonly changedTouches: InfiniteCanvasTouchList;
    readonly touches: InfiniteCanvasTouchList;
}

export * from './infinite-canvas-touch-list';