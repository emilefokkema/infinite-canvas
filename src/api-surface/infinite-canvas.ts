import {Config} from "./config";
import { Units } from "./units";
import {InfiniteCanvasRenderingContext2D} from "./infinite-canvas-rendering-context-2d";
import {EventMap} from "./event-map";
import { TransformationEvent } from "./transformation-event";
import { Transformed } from './transformed';

export interface InfiniteCanvasEventHandlers extends DocumentAndElementEventHandlers, GlobalEventHandlers {
    /**
     * The [event handler property](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Building_blocks/Events#event_handler_properties) for the {@link EventMap.transformationstart} event
     */
    ontransformationstart: ((this: InfiniteCanvas, event: TransformationEvent) => any) | null;
    /**
     * The [event handler property](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Building_blocks/Events#event_handler_properties) for the {@link EventMap.transformationchange} event
     */
    ontransformationchange: ((this: InfiniteCanvas, event: TransformationEvent) => any) | null;
    /**
     * The [event handler property](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Building_blocks/Events#event_handler_properties) for the {@link EventMap.transformationend} event
     */
    ontransformationend: ((this: InfiniteCanvas, event: TransformationEvent) => any) | null;
    /**
     * The [event handler property](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Building_blocks/Events#event_handler_properties) for the {@link EventMap.draw} event
     */
    ondraw: ((this: InfiniteCanvas, event: TransformationEvent) => any) | null;
    /**
     * The [event handler property](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Building_blocks/Events#event_handler_properties) for the {@link EventMap.wheelignored} event
     */
    onwheelignored:  ((this: InfiniteCanvas, event: Event) => any) | null;
    /**
     * The [event handler property](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Building_blocks/Events#event_handler_properties) for the {@link EventMap.touchignored} event
     */
    ontouchignored:  ((this: InfiniteCanvas, event: Event) => any) | null;
    /**
     * See [`addEventListener`](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener) 
     */
    addEventListener<K extends keyof EventMap>(type: K, listener: (this: InfiniteCanvas, ev: EventMap[K]) => any, options?: boolean | AddEventListenerOptions): void;
    addEventListener(type: string, listener: EventListenerOrEventListenerObject, options?: boolean | AddEventListenerOptions): void;
    /**
     * See [`removeEventListener`](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/removeEventListener)
     */
    removeEventListener<K extends keyof EventMap>(type: K, listener: (this: InfiniteCanvas, ev: EventMap[K]) => any, options?: boolean | EventListenerOptions): void;
    removeEventListener(type: string, listener: EventListenerOrEventListenerObject, options?: boolean | EventListenerOptions): void;
}

export interface InfiniteCanvas extends Config, InfiniteCanvasEventHandlers, Transformed{
    /**
     * This methods return the {@link InfiniteCanvasRenderingContext2D} belonging to this instance of {@link InfiniteCanvas}
     * 
     * @param contextType for (partial) compatibility with the other [getContext()](https://developer.mozilla.org/en-US/docs/Web/API/HTMLCanvasElement/getContext)
     */
    getContext(contextType?: '2d'): InfiniteCanvasRenderingContext2D;
}

export interface InfiniteCanvasCtr {
    /**
     * Creates a new instance of an {@link InfiniteCanvas}
     * 
     * @param canvas The canvas on which to draw the {@link InfiniteCanvas}
     * @param config The {@link InfiniteCanvas}'s initial configuration
     */
    new(canvas: HTMLCanvasElement, config?: Partial<Config>): InfiniteCanvas;
    prototype: InfiniteCanvas;
    /**
     * Static property with a value of {@link Units.CANVAS}
     */
    CANVAS_UNITS: Units;
    /**
     * Static property with a value of {@link Units.CSS}
     */
    CSS_UNITS: Units;
};

declare var InfiniteCanvasConstructor: InfiniteCanvasCtr;

export * from './config'
export * from './event-map'
export * from './infinite-canvas-rendering-context-2d'
export * from './units'
export default InfiniteCanvasConstructor;
