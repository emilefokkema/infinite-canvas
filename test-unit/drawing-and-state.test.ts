import { beforeEach, describe, expect, it} from 'vitest'
import {CanvasContextMock} from "./canvas-context-mock";
import {InfiniteCanvasRenderingContext2D} from "../src/api-surface/infinite-canvas-rendering-context-2d";
import { createInfiniteCanvasTestFixture } from './infinite-canvas-test-fixture';

describe('an infinite canvas context', () => {

	let infiniteContext: InfiniteCanvasRenderingContext2D;
	let contextMock: CanvasContextMock;


    beforeEach(() => {
		({contextMock, infiniteContext} = createInfiniteCanvasTestFixture());
	});

    describe("whose state is changed and who draws something", () => {
		let red: string;
		let blue: string;

		beforeEach(() => {
			red = "#f00";
			blue = "#00f";
			infiniteContext.fillStyle = red;
			infiniteContext.fillRect(1, 1, 2, 2);
		});

		it("should have modified the context correctly", () => {
			expect(contextMock.getLog()).toMatchSnapshot();
		});

		describe("and who changes state, draws something, changes state back and draws something again", () => {

			beforeEach(() => {
				infiniteContext.fillStyle = blue;
				infiniteContext.fillRect(5, 1, 2, 2);
				infiniteContext.fillStyle = red;
				contextMock.clear();
				infiniteContext.fillRect(9, 1, 2, 2);
			});

			it("should have set a new state three times", () => {
				expect(contextMock.getLog()).toMatchSnapshot();
			});

			describe("and who then clears a rect containing the second drawing", () => {

				beforeEach(() => {
					contextMock.clear();
					infiniteContext.clearRect(5, 1, 2, 2);
				});

				it("should only have set the remaining state", () => {
					expect(contextMock.getLog()).toMatchSnapshot();
				});
			});
		});

		describe("and which then changes the state and clears part of the drawing and draws something else", () => {

			beforeEach(() => {
				infiniteContext.fillStyle = blue;
				infiniteContext.clearRect(2, 0, 4, 4);
				contextMock.clear();
				infiniteContext.fillRect(3, 1, 1, 1);
			});

			it("should have drawn using the state from before the clearing", () => {
				expect(contextMock.getLog()).toMatchSnapshot();
			});
		});

		describe("and which then clears an area containing that instruction", () => {

			beforeEach(() => {
				contextMock.clear();
				infiniteContext.clearRect(0, 0, 4, 4);
			});

			it("should have cleared a rectangle", () => {
				expect(contextMock.getLog()).toMatchSnapshot();
			});

			describe("and which then draws something else without changing the state", () => {

				beforeEach(() => {
					contextMock.clear();
					infiniteContext.fillRect(1, 1, 2, 2);
				});

				it("should have cleared a rect only once more and should still use the old state", () => {
					expect(contextMock.getLog()).toMatchSnapshot();
				});
			});

			describe("and which then draws something else", () => {

				beforeEach(() => {
					infiniteContext.fillStyle = blue;
					contextMock.clear();
					infiniteContext.fillRect(1, 1, 2, 2);
				});

				it("should have cleared a rect only once more and should not have executed the old instruction again", () => {
					expect(contextMock.getLog()).toMatchSnapshot();
				});
			});
		});

		describe("and which then draws something else", () => {

			beforeEach(() => {
				contextMock.clear();
				infiniteContext.fillRect(4, 1, 2, 2);
			});

			it("should not have altered the state", () => {
				expect(contextMock.getLog()).toMatchSnapshot();
			});

			describe("and which then clears the first part", () => {

				beforeEach(() => {
					contextMock.clear();
					infiniteContext.clearRect(0, 0, 3.5, 4);
				});

				it("should have remembered the state for the second part", () => {
					expect(contextMock.getLog()).toMatchSnapshot();
				});
			});

			describe("and which then clears the first part and part of the second", () => {

				beforeEach(() => {
					contextMock.clear();
					infiniteContext.clearRect(0, 0, 4.5, 4);
				});

				it("should have remembered the state for the second part", () => {
					expect(contextMock.getLog()).toMatchSnapshot();
				});
			});
		});
	});

    describe("that saves state, begins a path, restores state and fills a rect", () => {

		beforeEach(() => {
			infiniteContext.save();
			infiniteContext.beginPath();
			infiniteContext.moveTo(1, 1);
			infiniteContext.restore();
			contextMock.clear();
			infiniteContext.fillRect(0, 0, 1, 1);
		});

		it("should end up with an equal number of saves and restores", () => {
			expect(contextMock.getLog()).toMatchSnapshot();
		});
	});

    describe("that saves, creates a path, fills two rects, restores and strokes", () => {

		beforeEach(() => {
			infiniteContext.save();
			infiniteContext.translate(1, 1);
			infiniteContext.beginPath();
			infiniteContext.rect(0, 0, 1, 1);
			infiniteContext.save();
			infiniteContext.fillRect(0, 0, 1, 1);
			infiniteContext.fillRect(0, 0, 1, 1);
			infiniteContext.restore();
			infiniteContext.stroke();
			infiniteContext.restore();
			contextMock.clear();
			infiniteContext.fillRect(0, 0, 1, 1)
		});

		it("should have the same number of saves and restores", () => {
			expect(contextMock.getLog()).toMatchSnapshot();
		});
	});

    describe("that saves, begins a path, begins another path, fills it, restores and then fills a rect", () => {

		beforeEach(() => {
			infiniteContext.save();
			infiniteContext.beginPath();
			infiniteContext.moveTo(0, 0);
			infiniteContext.beginPath();
			infiniteContext.moveTo(0, 0);
			infiniteContext.lineTo(2, 0);
			infiniteContext.lineTo(2, 2);
			infiniteContext.fill();
			infiniteContext.restore();
			contextMock.clear();
			infiniteContext.fillRect(0, 0, 1, 1);
		});

		it("should end up with an equal number of saves and restores", () => {
			expect(contextMock.getLog()).toMatchSnapshot();
		});
	});

    describe("that does this", () => {

		beforeEach(() => {
			infiniteContext.save();
			infiniteContext.fillStyle = '#f00';
			infiniteContext.translate(10, 10);
			infiniteContext.fillRect(0, 0, 25, 25);
			infiniteContext.restore();
			infiniteContext.save();
			infiniteContext.fillStyle = '#f00';
			infiniteContext.translate(60, 10);
			contextMock.clear();
			infiniteContext.fillRect(0, 0, 25, 25);
		});

		it("should have done this", () => {
			expect(contextMock.getLog()).toMatchSnapshot();
		});
	});

    describe("that draws a path, transforms and then strokes", () => {

		beforeEach(() => {
			infiniteContext.beginPath();
			infiniteContext.rect(10, 10, 100, 100);
			infiniteContext.lineWidth = 6;
			infiniteContext.transform(.2, 0, 0, 1, 0, 0);
			contextMock.clear();
			infiniteContext.stroke();
		});

		it("should have kept that order", () => {
			expect(contextMock.getLog()).toMatchSnapshot();
		});
	});

    describe("that draws a square by translating", () => {

		beforeEach(() => {
			infiniteContext.beginPath();
			infiniteContext.translate(1, 0);
			infiniteContext.moveTo(0,0);
			infiniteContext.translate(0, 1);
			infiniteContext.lineTo(0,0);
			infiniteContext.translate(-1,0);
			infiniteContext.lineTo(0,0);
			infiniteContext.translate(0, -1);
			infiniteContext.lineTo(0,0);
			contextMock.clear();
			infiniteContext.fill();
		});

		it("should have called transform before each move", () => {
			expect(contextMock.getLog()).toMatchSnapshot();
		});

		describe("and then partly clears the drawn square", () => {

			beforeEach(() => {
				contextMock.clear();
				infiniteContext.clearRect(0.5, 0, 2, 2);
			});

			it("should have added a clearRect", () => {
				expect(contextMock.getLog()).toMatchSnapshot();
			});
		});

		describe("and then fully clears the drawn square", () => {

			beforeEach(() => {
				contextMock.clear();
				infiniteContext.clearRect(-1, -1, 3, 3);
			});

			it("should not have added a clearRect", () => {
				expect(contextMock.getLog()).toMatchSnapshot();
			});
		});
	});

    describe('that saves and fills a rect', () => {

		beforeEach(() => {
			infiniteContext.save();
			infiniteContext.fillRect(10, 10, 20, 20)
		});

		it("should call restore() after drawing", () => {
			expect(contextMock.getLog()).toMatchSnapshot();
		});
	})

    describe("saves, changes state, begins drawing a path", () => {

		beforeEach(() => {
			infiniteContext.fillStyle = "#f00";
			infiniteContext.save();
			infiniteContext.fillStyle = "#00f";
			infiniteContext.beginPath();
			infiniteContext.rect(0, 0, 10, 10);
			
		});

		describe("and then restores to the previous state, fills a rect and then fills the path", () => {

			beforeEach(() => {
				infiniteContext.restore();
				infiniteContext.fillRect(10, 10, 20, 20);
				contextMock.clear();
				infiniteContext.fill();
			});

			it("should end up with an equal number of saves and restores", () => {
				expect(contextMock.getLog()).toMatchSnapshot();
			});
		});

		describe("and then fills a rect, restores to the previous state and then fills the path", () => {

			beforeEach(() => {
				infiniteContext.fillRect(10, 10, 20, 20);
				infiniteContext.restore();
				contextMock.clear();
				infiniteContext.fill();
			});

			it("should end up with an equal number of saves and restores", () => {
				expect(contextMock.getLog()).toMatchSnapshot();
			});
		});
	});


})