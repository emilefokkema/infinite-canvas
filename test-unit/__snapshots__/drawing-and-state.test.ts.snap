// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`an infinite canvas context > saves, changes state, begins drawing a path > and then fills a rect, restores to the previous state and then fills the path > should end up with an equal number of saves and restores 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.fillStyle = "#f00"",
  "context.save()",
  "context.fillStyle = "#00f"",
  "context.lineWidth = 1",
  "context.beginPath()",
  "context.moveTo(10,10)",
  "context.lineTo(30,10)",
  "context.lineTo(30,30)",
  "context.lineTo(10,30)",
  "context.closePath()",
  "context.moveTo(10,10)",
  "context.fill()",
  "context.lineWidth = 1",
  "context.beginPath()",
  "context.moveTo(0,0)",
  "context.lineTo(10,0)",
  "context.lineTo(10,10)",
  "context.lineTo(0,10)",
  "context.closePath()",
  "context.moveTo(0,0)",
  "context.restore()",
  "context.lineWidth = 1",
  "context.fill()",
]
`;

exports[`an infinite canvas context > saves, changes state, begins drawing a path > and then restores to the previous state, fills a rect and then fills the path > should end up with an equal number of saves and restores 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.fillStyle = "#f00"",
  "context.lineWidth = 1",
  "context.beginPath()",
  "context.moveTo(10,10)",
  "context.lineTo(30,10)",
  "context.lineTo(30,30)",
  "context.lineTo(10,30)",
  "context.closePath()",
  "context.moveTo(10,10)",
  "context.fill()",
  "context.lineWidth = 1",
  "context.save()",
  "context.fillStyle = "#00f"",
  "context.beginPath()",
  "context.moveTo(0,0)",
  "context.lineTo(10,0)",
  "context.lineTo(10,10)",
  "context.lineTo(0,10)",
  "context.closePath()",
  "context.moveTo(0,0)",
  "context.restore()",
  "context.lineWidth = 1",
  "context.fill()",
]
`;

exports[`an infinite canvas context > that does this > should have done this 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.save()",
  "context.fillStyle = "#f00"",
  "context.lineWidth = 1",
  "context.setTransform(1,0,0,1,10,10)",
  "context.beginPath()",
  "context.moveTo(0,0)",
  "context.lineTo(25,0)",
  "context.lineTo(25,25)",
  "context.lineTo(0,25)",
  "context.closePath()",
  "context.moveTo(0,0)",
  "context.fill()",
  "context.setTransform(1,0,0,1,60,10)",
  "context.beginPath()",
  "context.moveTo(0,0)",
  "context.lineTo(25,0)",
  "context.lineTo(25,25)",
  "context.lineTo(0,25)",
  "context.closePath()",
  "context.moveTo(0,0)",
  "context.fill()",
  "context.restore()",
]
`;

exports[`an infinite canvas context > that draws a path, transforms and then strokes > should have kept that order 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.beginPath()",
  "context.moveTo(10,10)",
  "context.lineTo(110,10)",
  "context.lineTo(110,110)",
  "context.lineTo(10,110)",
  "context.closePath()",
  "context.moveTo(10,10)",
  "context.lineWidth = 6",
  "context.setTransform(0.2,0,0,1,0,0)",
  "context.stroke()",
]
`;

exports[`an infinite canvas context > that draws a square by translating > and then fully clears the drawn square > should not have added a clearRect 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
]
`;

exports[`an infinite canvas context > that draws a square by translating > and then partly clears the drawn square > should have added a clearRect 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.beginPath()",
  "context.setTransform(1,0,0,1,1,0)",
  "context.moveTo(0,0)",
  "context.setTransform(1,0,0,1,1,1)",
  "context.lineTo(0,0)",
  "context.setTransform(1,0,0,1,0,1)",
  "context.lineTo(0,0)",
  "context.setTransform(1,0,0,1,0,0)",
  "context.lineTo(0,0)",
  "context.lineWidth = 1",
  "context.fill()",
  "context.save()",
  "context.transform(1,0,0,1,0,0)",
  "context.clearRect(0.5,0,2,2)",
  "context.restore()",
]
`;

exports[`an infinite canvas context > that draws a square by translating > should have called transform before each move 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.beginPath()",
  "context.setTransform(1,0,0,1,1,0)",
  "context.moveTo(0,0)",
  "context.setTransform(1,0,0,1,1,1)",
  "context.lineTo(0,0)",
  "context.setTransform(1,0,0,1,0,1)",
  "context.lineTo(0,0)",
  "context.setTransform(1,0,0,1,0,0)",
  "context.lineTo(0,0)",
  "context.lineWidth = 1",
  "context.fill()",
]
`;

exports[`an infinite canvas context > that saves and fills a rect > should call restore() after drawing 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.save()",
  "context.lineWidth = 1",
  "context.beginPath()",
  "context.moveTo(10,10)",
  "context.lineTo(30,10)",
  "context.lineTo(30,30)",
  "context.lineTo(10,30)",
  "context.closePath()",
  "context.moveTo(10,10)",
  "context.fill()",
  "context.restore()",
]
`;

exports[`an infinite canvas context > that saves state, begins a path, restores state and fills a rect > should end up with an equal number of saves and restores 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.lineWidth = 1",
  "context.beginPath()",
  "context.moveTo(0,0)",
  "context.lineTo(1,0)",
  "context.lineTo(1,1)",
  "context.lineTo(0,1)",
  "context.closePath()",
  "context.moveTo(0,0)",
  "context.fill()",
]
`;

exports[`an infinite canvas context > that saves, begins a path, begins another path, fills it, restores and then fills a rect > should end up with an equal number of saves and restores 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.save()",
  "context.beginPath()",
  "context.moveTo(0,0)",
  "context.lineTo(2,0)",
  "context.lineTo(2,2)",
  "context.lineWidth = 1",
  "context.fill()",
  "context.restore()",
  "context.lineWidth = 1",
  "context.beginPath()",
  "context.moveTo(0,0)",
  "context.lineTo(1,0)",
  "context.lineTo(1,1)",
  "context.lineTo(0,1)",
  "context.closePath()",
  "context.moveTo(0,0)",
  "context.fill()",
]
`;

exports[`an infinite canvas context > that saves, creates a path, fills two rects, restores and strokes > should have the same number of saves and restores 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.save()",
  "context.setTransform(1,0,0,1,1,1)",
  "context.save()",
  "context.lineWidth = 1",
  "context.beginPath()",
  "context.moveTo(0,0)",
  "context.lineTo(1,0)",
  "context.lineTo(1,1)",
  "context.lineTo(0,1)",
  "context.closePath()",
  "context.moveTo(0,0)",
  "context.fill()",
  "context.beginPath()",
  "context.moveTo(0,0)",
  "context.lineTo(1,0)",
  "context.lineTo(1,1)",
  "context.lineTo(0,1)",
  "context.closePath()",
  "context.moveTo(0,0)",
  "context.fill()",
  "context.restore()",
  "context.beginPath()",
  "context.moveTo(0,0)",
  "context.lineTo(1,0)",
  "context.lineTo(1,1)",
  "context.lineTo(0,1)",
  "context.closePath()",
  "context.moveTo(0,0)",
  "context.lineWidth = 1",
  "context.stroke()",
  "context.restore()",
  "context.lineWidth = 1",
  "context.beginPath()",
  "context.moveTo(0,0)",
  "context.lineTo(1,0)",
  "context.lineTo(1,1)",
  "context.lineTo(0,1)",
  "context.closePath()",
  "context.moveTo(0,0)",
  "context.fill()",
]
`;

exports[`an infinite canvas context > whose state is changed and who draws something > and which then changes the state and clears part of the drawing and draws something else > should have drawn using the state from before the clearing 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.fillStyle = "#f00"",
  "context.lineWidth = 1",
  "context.beginPath()",
  "context.moveTo(1,1)",
  "context.lineTo(3,1)",
  "context.lineTo(3,3)",
  "context.lineTo(1,3)",
  "context.closePath()",
  "context.moveTo(1,1)",
  "context.fill()",
  "context.fillStyle = "#00f"",
  "context.save()",
  "context.transform(1,0,0,1,0,0)",
  "context.clearRect(2,0,4,4)",
  "context.restore()",
  "context.beginPath()",
  "context.moveTo(3,1)",
  "context.lineTo(4,1)",
  "context.lineTo(4,2)",
  "context.lineTo(3,2)",
  "context.closePath()",
  "context.moveTo(3,1)",
  "context.fill()",
]
`;

exports[`an infinite canvas context > whose state is changed and who draws something > and which then clears an area containing that instruction > and which then draws something else > should have cleared a rect only once more and should not have executed the old instruction again 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.fillStyle = "#00f"",
  "context.lineWidth = 1",
  "context.beginPath()",
  "context.moveTo(1,1)",
  "context.lineTo(3,1)",
  "context.lineTo(3,3)",
  "context.lineTo(1,3)",
  "context.closePath()",
  "context.moveTo(1,1)",
  "context.fill()",
]
`;

exports[`an infinite canvas context > whose state is changed and who draws something > and which then clears an area containing that instruction > and which then draws something else without changing the state > should have cleared a rect only once more and should still use the old state 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.fillStyle = "#f00"",
  "context.lineWidth = 1",
  "context.beginPath()",
  "context.moveTo(1,1)",
  "context.lineTo(3,1)",
  "context.lineTo(3,3)",
  "context.lineTo(1,3)",
  "context.closePath()",
  "context.moveTo(1,1)",
  "context.fill()",
]
`;

exports[`an infinite canvas context > whose state is changed and who draws something > and which then clears an area containing that instruction > should have cleared a rectangle 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
]
`;

exports[`an infinite canvas context > whose state is changed and who draws something > and which then draws something else > and which then clears the first part > should have remembered the state for the second part 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.fillStyle = "#f00"",
  "context.lineWidth = 1",
  "context.beginPath()",
  "context.moveTo(4,1)",
  "context.lineTo(6,1)",
  "context.lineTo(6,3)",
  "context.lineTo(4,3)",
  "context.closePath()",
  "context.moveTo(4,1)",
  "context.fill()",
]
`;

exports[`an infinite canvas context > whose state is changed and who draws something > and which then draws something else > and which then clears the first part and part of the second > should have remembered the state for the second part 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.fillStyle = "#f00"",
  "context.lineWidth = 1",
  "context.beginPath()",
  "context.moveTo(4,1)",
  "context.lineTo(6,1)",
  "context.lineTo(6,3)",
  "context.lineTo(4,3)",
  "context.closePath()",
  "context.moveTo(4,1)",
  "context.fill()",
  "context.save()",
  "context.transform(1,0,0,1,0,0)",
  "context.clearRect(0,0,4.5,4)",
  "context.restore()",
]
`;

exports[`an infinite canvas context > whose state is changed and who draws something > and which then draws something else > should not have altered the state 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.fillStyle = "#f00"",
  "context.lineWidth = 1",
  "context.beginPath()",
  "context.moveTo(1,1)",
  "context.lineTo(3,1)",
  "context.lineTo(3,3)",
  "context.lineTo(1,3)",
  "context.closePath()",
  "context.moveTo(1,1)",
  "context.fill()",
  "context.beginPath()",
  "context.moveTo(4,1)",
  "context.lineTo(6,1)",
  "context.lineTo(6,3)",
  "context.lineTo(4,3)",
  "context.closePath()",
  "context.moveTo(4,1)",
  "context.fill()",
]
`;

exports[`an infinite canvas context > whose state is changed and who draws something > and who changes state, draws something, changes state back and draws something again > and who then clears a rect containing the second drawing > should only have set the remaining state 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.fillStyle = "#f00"",
  "context.lineWidth = 1",
  "context.beginPath()",
  "context.moveTo(1,1)",
  "context.lineTo(3,1)",
  "context.lineTo(3,3)",
  "context.lineTo(1,3)",
  "context.closePath()",
  "context.moveTo(1,1)",
  "context.fill()",
  "context.beginPath()",
  "context.moveTo(9,1)",
  "context.lineTo(11,1)",
  "context.lineTo(11,3)",
  "context.lineTo(9,3)",
  "context.closePath()",
  "context.moveTo(9,1)",
  "context.fill()",
]
`;

exports[`an infinite canvas context > whose state is changed and who draws something > and who changes state, draws something, changes state back and draws something again > should have set a new state three times 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.fillStyle = "#f00"",
  "context.lineWidth = 1",
  "context.beginPath()",
  "context.moveTo(1,1)",
  "context.lineTo(3,1)",
  "context.lineTo(3,3)",
  "context.lineTo(1,3)",
  "context.closePath()",
  "context.moveTo(1,1)",
  "context.fill()",
  "context.fillStyle = "#00f"",
  "context.beginPath()",
  "context.moveTo(5,1)",
  "context.lineTo(7,1)",
  "context.lineTo(7,3)",
  "context.lineTo(5,3)",
  "context.closePath()",
  "context.moveTo(5,1)",
  "context.fill()",
  "context.fillStyle = "#f00"",
  "context.beginPath()",
  "context.moveTo(9,1)",
  "context.lineTo(11,1)",
  "context.lineTo(11,3)",
  "context.lineTo(9,3)",
  "context.closePath()",
  "context.moveTo(9,1)",
  "context.fill()",
]
`;

exports[`an infinite canvas context > whose state is changed and who draws something > should have modified the context correctly 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.fillStyle = "#f00"",
  "context.lineWidth = 1",
  "context.beginPath()",
  "context.moveTo(1,1)",
  "context.lineTo(3,1)",
  "context.lineTo(3,3)",
  "context.lineTo(1,3)",
  "context.closePath()",
  "context.moveTo(1,1)",
  "context.fill()",
]
`;
